numba/cuda/cudadrv/devicearray.py

```diff
@@ -116,10 +116,20 @@ def __init__(self, shape, strides, dtype, stream=0, writeback=None,
 
     @property
     def __cuda_array_interface__(self):
+        if self.device_ctypes_pointer.value is not None:
+            ptr = self.device_ctypes_pointer.value
+        else:
+            ptr = 0
```

Comment:
I think we should document that `ptr` can be `0`, and what that means.

Reply:
Or rather document that the `data` may include a value `0` as first element.

Reply:
@pentschev Thanks for quick reply! How about this:
```diff
diff --git a/docs/source/cuda/cuda_array_interface.rst b/docs/source/cuda/cuda_array_interface.rst
index fa14d41..1aa3e95 100644
--- a/docs/source/cuda/cuda_array_interface.rst
+++ b/docs/source/cuda/cuda_array_interface.rst
@@ -34,8 +34,9 @@ contain the following entries:
 
 - **data**: `(integer, boolean)`
 
-    The **data** is a 2-tuple.  The first element data pointer
+    The **data** is a 2-tuple.  The first element is the data pointer
     as a Python `int` (or `long`).  The data must be device-accessible.
+    For zero-size arrays, use `0` here.
     The second element is the read-only flag as a Python `bool`.
 
     Because the user of the interface may or may not be in the same context,
```

Reply:
@sklam comment for the diff above?

Reply:
@leofang, thanks for pointing this out. I didn't pay enough attention to the discussion of making `strides` mandatory. I am a bit worried about this because numpy has this as optional with the default set to `None`. A consumer can retrieve the strides as `__cuda_array_interface__.get("strides")`, which will return `None` if the key is absent or return the actual value for it. I would prefer to leave it as optional.

The zero-sized array part looks good.

Reply:
@sklam, the NumPy doc is actually wrong. The NumPy implementation actually always has `strides` in place, and set to `None` for C-contigus arrays:
```python
>>> import numpy as np
>>> a = np.arange(10)
>>> a.flags['C_CONTIGUOUS']
True
>>> a.__array_interface__['strides'] == None
True
>>> np.__version__
'1.17.2'
```
I can raise an issue in NumPy to ask them to fix it. But having a mandatory `strides` really helps us here.

cc: @pentschev

Reply:
relevant lines in NumPy:
https://github.com/numpy/numpy/blob/7214ca4688545b432c45287195e2f46c5e418ce8/numpy/core/src/multiarray/getset.c#L239-L246

Reply:
I interpret it as consumers should not rely on the "strides" to exist, but producers can choose to always fill it in.

Reply:
As a consumer I'd also like to be able to retrieve it as `__cuda_array_interface__['strides']` without checking the presence of the key. We're gonna check if `strides is None` later anyway.

Reply:
in python, you can do it in 1-line with `dict.get()`, i.e. `__cuda_array_interface__.get('strides')`, it doesn't raise.

Reply:
> But having a mandatory `strides` really helps us here.

...Sorry for bombing you with multiple messages. The "help" is that we can ensure the array is C contiguous. @pentschev has suggested adding a `'c_contiguous'` field to the interface, but having this specified would do both jobs at once.

Like current Numba, PyTorch also always explicitly specifies `strides=(...)` even if it's C contiguous. I'd like to take an amended protocol to them and "strong-arm" them to make a fix 😄

Reply:
Oh I didn't refresh and just saw your posts now.

Reply:
@sklam are you suggesting a missing `strides` should be used to for representing C contiguity? I wouldn't mind doing this, but I need everyone on board. 

cc: @pentschev

Reply:
Yes, the numpy doc say that the default is None. Thus implying c contiguous if strides is missing.

Reply:
> Yes, the numpy doc say that the default is None. Thus implying c contiguous if strides is missing.

No @sklam, I think you misunderstood my reply. "Default is None" is different from "strides is missing". The former suggests that `'strides'` is always there (with value `None`), while the latter says literally `'strides'` is not there. 

Btw, see how easy we can parse the protocol differently? It'd be really, really helpful if we can say our intent loud and clear in the standard.

Reply:
And let me refer to my earlier comment again:
https://github.com/numba/numba/pull/4609#discussion_r328772429

Reply:
To me, "Default is None" means "None if not given".  I think this interpretation is common in numpy docs.

For instance, for `np.asarray`:

> order : {‘C’, ‘F’}, **optional**
> Whether to use row-major (C-style) or column-major (Fortran-style) memory representation. **Defaults to ‘C’.**

(from https://docs.scipy.org/doc/numpy/reference/generated/numpy.asarray.html?highlight=asarray#numpy.asarray :)

If `order` is not provided, default it to 'C'.

In Python doc, it also has the following for `dict.get`:

> get(key[, default])¶
> Return the value for key if key is in the dictionary, else default. **If default is not given, it defaults to None**, so that this method never raises a KeyError.

(from https://docs.python.org/3/library/stdtypes.html#dict.get)

In the array interface doc (https://docs.scipy.org/doc/numpy/reference/arrays.interface.html), it has:

> strides **(optional)**
> Either None to indicate a C-style contiguous array or a Tuple of strides which... **The default is None** which implies a C-style contiguous memory buffer...
> **Default: None (C-style contiguous)**

Reply:
I see that cupy is taking the similar interpretation:

https://github.com/cupy/cupy/blob/36ecb4423762a79aea98bd1bbb40a3903d28ac18/cupy/core/core.pyx#L2748-L2755

Reply:
@sklam I am not convinced. The comparison to `np.asarray` is apple to orange (it's just Python's default argument), and as I pointed out the NumPy implementation does not follow your interpretation. 

So, let me ask this instead: Do you insist on your interpretation in order to move on? I am fine either way, just wonder if you're still open to discussion or you're making a call. 

Thanks.

Reply:
@leofang Maybe I'm missing something, but if we clarify that either missing `strides` or `strides=None` means C-contiguous, then consumer code just need to do:
```
strides = exporter.__cuda_array_interface__.get('strides')
if strides is None:
   ... # the buffer is C-contiguous
else:
    ... # handle strides if possible else raise exception
```

About NumPy, the docs are rather clear to me, `strides` may be missing. But let's look more carefully at the implementation.

The producer code always set `strides`, and use a `None` value to signal C contiguity. IOW, although `strides` is optional and may be missing, they decided to always set it.
https://github.com/numpy/numpy/blob/7214ca4688545b432c45287195e2f46c5e418ce8/numpy/core/src/multiarray/getset.c#L295-L297

The consumer code is in in `PyArray_FromInterface()`, please note the check `if (attr != NULL && attr != Py_None)`, that means that `strides` may be missing.
https://github.com/numpy/numpy/blob/7214ca4688545b432c45287195e2f46c5e418ce8/numpy/core/src/multiarray/ctors.c#L2633-L2634

IMHO, both numba and cupy should follow NumPy specs and implementation as close as possible. That means:

- In the interface spec, `strides` may be missing. If `strides` is missing or it is present and the value is `None`, that means the buffer is C-contiguous.
- The producer code always set `strides`, using `None` to flag C-contiguous arrays. We need tests to ensure that the numba producer code always sets `strides`.
- The consumer code do not assume `strides` is present, as it is optional. We need special tests to check that the consumer code can handle missing strides, maybe using a fake exporter as we did in mpi4py's testsuite.

Reply:
Thanks @dalcinl for fresh thoughts! 

> clarify that either missing strides or strides=None means C-contiguous

We can do that.

> The producer code always set `strides`, ...
> The consumer code is ... `strides` may be missing.

I haven't thought about making a distinction in the protocol. I am a bit worried it'd be too lengthy, but I don't object for better clarity. 

Note that in the current PR we always set `strides` on the producer side.

Also note that it's not difficult to bring Numba and CuPy in sync. The CuPy counterpart is cupy/cupy#2491.

Reply:
> clarify that either missing strides or strides=None means C-contiguous

Yes, we should do that.

> I haven't thought about making a distinction between producer and consumer in the protocol. I am a bit worried it'd be too lengthy, but I don't object for better clarity.

If you worry about it being too long, I think the documentation should focus on the consumer perspective first. 

Please also note that we are close to cutting a release. I would like to make sure this gets in by EOD. I am happy to accept this PR if we can reverse the decision on making `strides` mandatory.  I see the patch for zero-sized array critical for many downstream projects.

Reply:
@sklam I'll try to work with you toward EOD! I'm on Gitter if you'd like to pin me.

How about this:
```diff
diff --git a/docs/source/cuda/cuda_array_interface.rst b/docs/source/cuda/cuda_array_interface.rst
index fa14d41..ba60472 100644
--- a/docs/source/cuda/cuda_array_interface.rst
+++ b/docs/source/cuda/cuda_array_interface.rst
@@ -49,14 +49,15 @@ contain the following entries:
     An integer for the version of the interface being exported.
     The current version is *2*.
 
-- **strides**: ``None`` or ``(integer, ...)``
 
-    If it is ``None``, the array is in C-contiguous layout. Otherwise, a tuple
-    of `int` (or `long`) is explicitly given for representing the number of
-    bytes to skip to access the next element at each dimension.
+The following are optional entries:
 
+- **strides**: ``None`` or ``(integer, ...)``
 
-The following are optional entries:
+    If **strides** is not given, or it is ``None``, the array is in
+    C-contiguous layout. Otherwise, a tuple of `int` (or `long`) is explicitly
+    given for representing the number of bytes to skip to access the next
+    element at each dimension.
 
 - **descr**
 ```
If the doc is OK, I'll proceed to modify the codes and tests. Thanks.

Reply:
just updated the diff, see above

Reply:
looks good

Reply:
@leofang My English skills are quite limited, but shouldn't it read `Versions 1 and 2 of the CUDA Array Interface ...` ?

Reply:
haha OK I'm on it. I followed the PR by @kkraus14

Reply:
Marked as resolved via 836cff5